### PR TITLE
Generic upgrade test and null_inserts test

### PR DESCRIPTION
**Pull request by roydahan**
## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by roydahan** [comment]
The following job should test the PR:
https://jenkins.scylladb.com/job/scylla-master/job/reproducers/job/rolling-upgrade-with-null-inserts/

Once PR is merged, we should copy the job to the correct jenkins folder and change the fork and branch.

**Comment by roydahan** [comment]
@Orenef11 once this reviewed and merged, you can use the same approach for alternator upgrade.

File: jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+// trick from https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/43
+def lib = library identifier: 'sct@snapshot', retriever: legacySCM(scm)
+
+rollingUpgradePipeline(
+    backend: 'gce',
+    base_versions: ['4.1'],
+
+    test_name: 'upgrade_test.UpgradeTest.test_generic_cluster_upgrade',
+    test_config: '''["test-cases/upgrades/generic-rolling-upgrade.yaml", "test-cases/upgrades/rolling-upgrade-with-null-inserts.yaml"]''',
```
Comment (roydahan) [inline jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile:11]: For some reason I have an issue here with the export, although I copied it from a similar test that using it (EaR longevity).

Comment (roydahan) [reply jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile]: export 'SCT_CONFIG_FILES=[test-cases/upgrades/generic-rolling-upgrade.yaml,' 'configurations/rolling-upgrade-with-null-inserts.yaml]'
03:15:44  + SCT_CONFIG_FILES='[test-cases/upgrades/generic-rolling-upgrade.yaml,'
03:15:44  /jenkins/slave/workspace/scylla-master/reproducers/rolling-upgrade-with-null-inserts/scylla-cluster-tests@tmp/durable-32eee7f7/script.sh: line 10: export: `configurations/rolling-upgrade-with-null-inserts.yaml]': not a valid identifier

Comment (roydahan) [reply jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile]: @fruch any idea why?

File: test-cases/upgrades/generic-rolling-upgrade.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# stress_during_entire_upgrade:       # To be configured in additional yaml
+# stress_after_cluster_upgrade:    # To be configured in additional yaml
+
+n_db_nodes: 3
+n_loaders: 1
+n_monitor_nodes: 1
+
+num_nodes_to_rollback: 2   # max is n_db_nodes -1, after upgrading 4 nodes last node can't be rollback
+
+linux_distro: 'centos'
```
Comment (fruch) [inline test-cases/upgrades/generic-rolling-upgrade.yaml:12]:
```suggestion
scylla_linux_distro: 'centos'
```

**Review by fruch — APPROVED** [review]
LGTM

File: jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile (hunk 1 above)
Comment (fruch) [reply jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile]: can you share the link ?

**Comment by fruch** [comment]
@roydahan looks good, one small typo on one of the yamls

File: jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile (hunk 1 above)
Comment (fruch) [reply jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile]: it's a bug in the `rollingUpgradePipeline.groovy` 

in other pipelines the test config is handled correctly:
```groovy
test_config = groovy.json.JsonOutput.toJson(params.test_config)
```

File: upgrade_test.py [hunk 3]
```diff
@@ -657,6 +659,83 @@ def test_rolling_upgrade(self):  # pylint: disable=too-many-locals,too-many-stat
...
+        prepare_thread_pool = self.run_stress_thread(stress_cmd=stress_before_upgrade)
+
+        InfoEvent('Waiting for stress_before_upgrade to finish')
+        self.verify_stress_thread(prepare_thread_pool)
+
+        # Starting workload during entire upgrade
+        InfoEvent('Starting stress_during_entire_upgrade workload')
+        stress_during_entire_upgrade = self.params.get('stress_during_entire_upgrade')
+        stress_thread_pool = self.run_stress_thread(stress_cmd=stress_during_entire_upgrade)
+        self.log.info('Sleeping for 60s to let cassandra-stress start before the rollback...')
```
Comment (Orenef11) [inline upgrade_test.py:682]: Please change `60` to `600`

File: upgrade_test.py [hunk 4]
```diff
@@ -657,6 +659,83 @@ def test_rolling_upgrade(self):  # pylint: disable=too-many-locals,too-many-stat
...
+        for i in range(nodes_num):
+            InfoEvent(f'Step{num_nodes_to_rollback * 2 + i + 1} - Upgrade node{i + 1}')
+            self.db_cluster.node_to_upgrade = self.db_cluster.nodes[indexes[i]]
+            self.log.info(f'Upgrade Node {self.db_cluster.node_to_upgrade.name} begins')
+            self.upgrade_node(self.db_cluster.node_to_upgrade, upgrade_sstables=upgrade_sstables)
+            self.log.info(f'Upgrade Node {self.db_cluster.node_to_upgrade.name} ended')
+            self.db_cluster.node_to_upgrade.check_node_health()
+            upgraded_nodes.append(self.db_cluster.node_to_upgrade)
+
+        InfoEvent(f'All nodes were upgraded successfully')
```
Comment (Orenef11) [inline upgrade_test.py:729]: Please remove the `f` format from the message.

File: upgrade_test.py [hunk 5]
```diff
@@ -657,6 +659,83 @@ def test_rolling_upgrade(self):  # pylint: disable=too-many-locals,too-many-stat
...
+
+        InfoEvent('Waiting for stress_during_entire_upgrade to finish')
```
Comment (Orenef11) [inline upgrade_test.py:731]: Why not combine the two events into one event?

File: upgrade_test.py [hunk 6]
```diff
@@ -657,6 +659,83 @@ def test_rolling_upgrade(self):  # pylint: disable=too-many-locals,too-many-stat
...
+        random.shuffle(indexes)
+
+        upgrade_sstables = self.params.get('upgrade_sstables')
+
+        if num_nodes_to_rollback > 0:
+            # Upgrade all nodes that should be rollback later
+            for i in range(num_nodes_to_rollback):
+                InfoEvent(f'Step{i + 1} - Upgrade node{i + 1}')
+                self.db_cluster.node_to_upgrade = self.db_cluster.nodes[indexes[i]]
+                self.log.info(f'Upgrade Node {self.db_cluster.node_to_upgrade.name} begins')
```
Comment (Orenef11) [inline upgrade_test.py:702]: Please not use f-format for log message

File: upgrade_test.py [hunk 7]
```diff
@@ -657,6 +659,83 @@ def test_rolling_upgrade(self):  # pylint: disable=too-many-locals,too-many-stat
...
+        # prepare an array containing the indexes
+        indexes = list(range(nodes_num))
+        # shuffle it so we will upgrade the nodes in a random order
+        random.shuffle(indexes)
+
+        upgrade_sstables = self.params.get('upgrade_sstables')
+
+        if num_nodes_to_rollback > 0:
+            # Upgrade all nodes that should be rollback later
+            for i in range(num_nodes_to_rollback):
```
Comment (Orenef11) [inline upgrade_test.py:699]: Please change from `i` to `idx`

File: test-cases/upgrades/generic-rolling-upgrade.yaml (hunk 2 above)
Comment (roydahan) [reply test-cases/upgrades/generic-rolling-upgrade.yaml:12]: Didn't even know we have this.
What's the difference?
The original rolling-upgrade use linux_distro.

Comment (fruch) [reply test-cases/upgrades/generic-rolling-upgrade.yaml:12]: I always was called `scylla_linux_disto` but in the pipeline it was called linux_distro for some reason... I guess it was me being lazy...

File: upgrade_test.py (hunk 6 above)
Comment (roydahan) [reply upgrade_test.py:702]: Why?

File: upgrade_test.py (hunk 5 above)
Comment (roydahan) [reply upgrade_test.py:731]: Two separate logics.

File: upgrade_test.py (hunk 6 above)
Comment (Orenef11) [reply upgrade_test.py:702]: Because that:
https://docs.python.org/3/howto/logging.html#optimization

File: upgrade_test.py (hunk 3 above)
Comment (roydahan) [reply upgrade_test.py:682]: good catch, it's actually to change from 600 to 60.

File: upgrade_test.py (hunk 7 above)
Comment (roydahan) [reply upgrade_test.py:699]: For i is good enough and not less pythonic (or any other language) than idx.

File: upgrade_test.py (hunk 6 above)
Comment (roydahan) [reply upgrade_test.py:702]: Sorry, but none of these links convinced me.
This info log message is important to those that will try to debug an issue from the logs.

**Comment by roydahan** [comment]
Addressed some of @Orenef11 and one comment from @fruch.

File: jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,14 @@
+#!groovy
+
+// trick from https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/43
+def lib = library identifier: 'sct@snapshot', retriever: legacySCM(scm)
+
+rollingUpgradePipeline(
+    backend: 'gce',
+    base_versions: ['4.1'],
```
Comment (Orenef11) [inline jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile:8]: @roydahan 
Now that version `4.2` is official, it's not better to run on top of `4.2`

**Review by Orenef11 — COMMENTED** [review]
@roydahan 
LGTM

**Comment by fruch** [comment]
@roydahan do you have a working pipeline for this one ?

**Comment by fruch** [comment]
@roydahan also do we want this a triggered job ?

**Comment by roydahan** [comment]
> @roydahan do you have a working pipeline for this one ?

It's already here.

**Comment by roydahan** [comment]
> @roydahan also do we want this a triggered job ?

No

File: jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile (hunk 8 above)
Comment (roydahan) [reply jenkins-pipelines/rolling-upgrade-with-null-inserts.jenkinsfile:8]: Basicailly you're right now, but for now we can keep it 4.1 till later phases of 4.3.
In any case, I'll send another PR to have the base_versions as a job parameter for all rolling upgrades jobs.

**Pull request merged**